### PR TITLE
Select the correct conditional import when building dart dependencies.

### LINKED ISSUE #9413: Hot Restart fails for apps using packages with conditional imports
Hot Reload works fine.  Hot Restart crashes the app.  Hot Restart works fine in the `flutter create` template example on the same machine.

How do I go about getting more debug information for you?

```
[✓] Flutter (on Mac OS X 10.12.4 16E195, channel master)
    • Flutter at /Stuff/Projects/flutter
    • Framework revision ecdfe65828 (2 days ago), 2017-04-12 23:13:32 -0700
    • Engine revision bcc8779cb8
    • Tools Dart version 1.23.0-dev.11.6

[✓] Host Executable Compatibility
    • Downloaded executables execute on host

[✓] Android toolchain - develop for Android devices (Android SDK 22.0.1)
    • Android SDK at /Users/eseidel/Library/Android/sdk/
    • Platform android-N, build-tools 22.0.1
    • ANDROID_HOME = /Users/eseidel/Library/Android/sdk/
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.1, Build version 8E1000a
    • ios-deploy 1.9.1
    • CocoaPods version 1.2.0

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Gradle version 3.2
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.1.1)
    • Dart plugin version 171.4073.40
    • Flutter plugin version 12.1

[✓] Connected devices
    • iPhone 6 • EE230D8B-B327-4B27-9008-0290BC320663 • ios • iOS 10.3 (simulator)
```

### PR DESCRIPTION
**Pull request by sivachandra**
Fixes #9413.

For more info on conditional imports, see
https://github.com/munificent/dep-interface-libraries/blob/master/Proposal.md.

### CODE REVIEW

File: packages/flutter_tools/lib/src/dart/dependencies.dart [hunk 1]
```diff
@@ -7,6 +7,22 @@ import 'package:analyzer/analyzer.dart' as analyzer;
 import '../base/file_system.dart';
 import '../dart/package_map.dart';
 
+// List of flutter specific environment configurations.
+// See https://github.com/munificent/dep-interface-libraries/blob/master/Proposal.md
+final List<String> _configurationConstants = <String>['dart.library.io'];
```
Comment (sivachandra) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: We could potentially populate this list with all dart: libraries that flutter supports.

Comment (a-siva) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: Maybe  add a TODO or comment stating that we would have to populate configurationConstants with all the dart: libraries, potentially "dart.IsFlutterEngine" and account for strings in Platform.environment

**Review by a-siva — APPROVED** [review]

File: packages/flutter_tools/lib/src/dart/dependencies.dart [hunk 2]
```diff
@@ -27,10 +43,26 @@ class DartDependencySetBuilder {
...
+                uriAsString = configuration.uri.stringValue;
+                break;
+              }
+            }
+          }
+          if (uriAsString == null) {
+            final analyzer.UriBasedDirective uriBasedDirective = directive;
+            uriAsString = uriBasedDirective.uri.stringValue;
+          }
+        } else {
```
Comment (goderbauer) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: To make this slightly more readable and save one level of indention, could you turn this if arround:
```
if (foo)
   continue
do other stuff
```

File: packages/flutter_tools/lib/src/dart/dependencies.dart (hunk 1 above)
Comment (goderbauer) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: When/why should we add more?

Comment (a-siva) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: When flutter imports some package that uses any of the other strings

**Review by goderbauer — APPROVED** [review]

File: packages/flutter_tools/lib/src/dart/dependencies.dart [hunk 3]
```diff
@@ -7,6 +7,22 @@ import 'package:analyzer/analyzer.dart' as analyzer;
 import '../base/file_system.dart';
 import '../dart/package_map.dart';
 
+// List of flutter specific environment configurations.
+// See https://github.com/munificent/dep-interface-libraries/blob/master/Proposal.md
+final List<String> _configurationConstants = <String>['dart.library.io'];
+
+String _dottedNameToString(analyzer.DottedName dottedName) {
+  String result = '';
+  dottedName.components.forEach((identifier) {
```
Comment (goderbauer) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: nit: [AVOID using Iterable.forEach() with a function literal](https://www.dartlang.org/guides/language/effective-dart/usage#avoid-using-iterableforeach-with-a-function-literal)

**Pull request closed without merging**